File: aic-application/app/src/main/java/com/intel/gamepad/activity/GameDetailActivity.java
/* Copyright (C) 2021 Intel Corporation 
 * Licensed under the Apache License, Version 2.0 (the "License"); 
 * you may not use this file except in compliance with the License. 
 * You may obtain a copy of the License at
 *   
 *	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 * See the License for the specific language governing permissions and 
 * limitations under the License.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.intel.gamepad.activity;

import android.media.MediaCodecInfo;
import android.media.MediaCodecList;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;


import com.commonlibrary.utils.StatusBarUtil;
import com.commonlibrary.view.loadingDialog.LoadingDialog;
import com.google.android.material.button.MaterialButton;
import com.hjq.permissions.OnPermissionCallback;
import com.hjq.permissions.Permission;
import com.hjq.permissions.XXPermissions;
import com.intel.gamepad.R;
import com.intel.gamepad.app.AppConst;
import com.intel.gamepad.bean.GameListBean;
import com.intel.gamepad.controller.webrtc.BaseController;
import com.intel.gamepad.controller.webrtc.RTCControllerAndroid;
import com.intel.gamepad.owt.p2p.P2PHelper;
import com.intel.gamepad.utils.IPUtils;

import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class GameDetailActivity extends BaseActivity {
    private static final String PARAM_BEAN = "param_bean";
    /* This file needs to be placed in "/sdcard/" folder. */
    private final File codecWhitelistXMLFile = new File(Environment.getExternalStorageDirectory().getAbsoluteFile(),
            AppConst.CODEC_WHITELIST_FILENAME);
    private final String TAG = GameDetailActivity.class.toString();
    private GameListBean bean = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        StatusBarUtil.setTranslucentStatus(this);
        setContentView(R.layout.activity_game_detail);
        initView();
        loadData();
        EditText etServerIP = findViewById(R.id.etServerIP);
        EditText etCoturnIP = findViewById(R.id.etCoturnIP);
        EditText etPeerID = findViewById(R.id.etPeerID);
        EditText etClientID = findViewById(R.id.etClientID);
        CheckBox chkTest = findViewById(R.id.chkTest);
        if(etServerIP!=null){
            etServerIP.setText(IPUtils.loadIP());
            P2PHelper.serverIP = IPUtils.loadIP();
            etServerIP.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    if (s != null) {
                        IPUtils.saveip(s.toString());
                        P2PHelper.serverIP = IPUtils.loadIP();
                        if (s.toString().startsWith("http") && s.toString().contains(":")) {
                            List<String> ipInfo = IPUtils.getIp(s.toString());
                            if (ipInfo.size() > 0) {
                                etCoturnIP.setText(ipInfo.get(0));
                            }
                        }

                    }
                }

                @Override
                public void afterTextChanged(Editable s) {

                }
            });
        }

        if(etCoturnIP!=null){
            etCoturnIP.setText(IPUtils.loadCoturnIP());
            P2PHelper.strCoturn = IPUtils.loadCoturnIP();
            etCoturnIP.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    if (s != null) {
                        IPUtils.saveCoturn(s.toString());
                        P2PHelper.strCoturn = IPUtils.loadCoturnIP();
                    }
                }

                @Override
                public void afterTextChanged(Editable s) {

                }
            });

        }

        if(etPeerID!=null){
            etPeerID.setText(IPUtils.loadPeerID());
            P2PHelper.peerId = IPUtils.loadPeerID();
            etPeerID.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    if (s != null) {
                        IPUtils.savepeerid(s.toString());
                        P2PHelper.peerId = IPUtils.loadPeerID();
                    }
                }

                @Override
                public void afterTextChanged(Editable s) {

                }
            });
        }

        if(etClientID!=null){
            etClientID.setText(IPUtils.loadTokenID());
            P2PHelper.clientId = IPUtils.loadTokenID();
            etClientID.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    if (s != null) {
                        IPUtils.savetoken(s.toString());
                        P2PHelper.clientId = IPUtils.loadTokenID();
                    }
                }

                @Override
                public void afterTextChanged(Editable s) {

                }
            });
        }

        if(chkTest!=null){
            chkTest.setChecked(IPUtils.loadTest());
            chkTest.setOnCheckedChangeListener((buttonView, isChecked) -> IPUtils.savetest(isChecked));
        }

        checkOrientation();
    }

    private void checkOrientation() {
        Spinner orient_sp = findViewById(R.id.orient_sp);
        if(orient_sp!=null){
            List<String> orientList = new ArrayList<>();
            orientList.add(getResources().getString(R.string.landscape));
            orientList.add(getResources().getString(R.string.portrait));
            ArrayAdapter<String> codecAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, orientList);
            codecAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            orient_sp.setAdapter(codecAdapter);
            if (IPUtils.loadPortrait()) {
                orient_sp.setSelection(1);
            } else {
                orient_sp.setSelection(0);
            }
            orient_sp.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                @Override
                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                    IPUtils.savePortrait(position != 0);
                }

                @Override
                public void onNothingSelected(AdapterView<?> parent) {

                }
            });
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    private void initView() {
        initBackButton(R.id.ibtnBack);

        MaterialButton btnPlay = findViewById(R.id.btnPlay);
        if(btnPlay!=null){
            btnPlay.setOnClickListener(v -> {
                XXPermissions perm = XXPermissions.with(this)
                        .permission(Permission.CAMERA)
                        .permission(Permission.RECORD_AUDIO);
                if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.R){
                    perm.permission(Permission.MANAGE_EXTERNAL_STORAGE);
                }else{
                    perm.permission(Permission.Group.STORAGE);
                }
                perm.request(new OnPermissionCallback() {
                            @Override
                            public void onGranted(List<String> permissions, boolean all) {
                                if (all) {
                                    BaseController.manuallyPressBackButton.set(false);
                                    updateMediacodecXmlFile();
                                    requestStartGame();
                                }
                            }
                            @Override
                            public void onDenied(List<String> permissions, boolean never) {
                                if (never) {
                                    XXPermissions.startPermissionActivity(GameDetailActivity.this, permissions);
                                } else {
                                    for(String str:permissions)
                                    Toast.makeText(GameDetailActivity.this, "The "+str+" has not been granted", Toast.LENGTH_LONG).show();
                                }
                            }
                        });
            });
            btnPlay.requestFocus();
        }
    }


    private void requestStartGame() {
        LoadingDialog dlg = (LoadingDialog) LoadingDialog.show(this);
        gotoGamePlay();
        dlg.dismiss();
    }

    private void gotoGamePlay() {
        if (bean != null) {
            bean.setAddurl("android");
            PlayGameRtcActivity.actionStart(this, RTCControllerAndroid.NAME, bean.getIid(), bean.getConf());
        }
    }

    private void loadData() {
        bean = getIntent().getParcelableExtra(PARAM_BEAN);
        if (bean == null) {
            bean = new GameListBean();
            bean.setIid(1);
            bean.setConf("rts");
        }
    }

    private void updateMediacodecXmlFile() {
        String fileContent;
        String referenceXmlStr = prepareCodecWhitelistXmlStr();

        // file does not exist, create it
        if (!codecWhitelistXMLFile.exists()) {
            Log.d(TAG, "file not found, creating: " + codecWhitelistXMLFile.getName());

            try {
                FileUtils.fileWrite(codecWhitelistXMLFile.getAbsolutePath(), referenceXmlStr);
            } catch (Exception e) {
                Log.e(TAG, "error creating file: " + codecWhitelistXMLFile.getName());
                e.printStackTrace();
            }
            return;
        }

        // file exists, compare and replace content if required
        try {
            fileContent = FileUtils.fileRead(codecWhitelistXMLFile.getAbsolutePath());
        } catch (IOException e) {
            // error reading file, probably
            Log.e(TAG, "cannot update file (" + codecWhitelistXMLFile.getAbsolutePath()
                    + "), error reading file: " + e.getMessage());
            e.printStackTrace();
            return;
        }

        int match = fileContent.compareTo(referenceXmlStr);

        if (match != 0) {
            // file different than expected, replace content with expected one
            Log.d(TAG, "updating content of " + codecWhitelistXMLFile.getName());
            try {
                FileUtils.fileWrite(codecWhitelistXMLFile.getAbsolutePath(), referenceXmlStr);
            } catch (Exception e) {
                Log.d(TAG, "error updating file: " + codecWhitelistXMLFile.getName());
                e.printStackTrace();
            }
        } else {
            // do nothing, the expected file is already present.
            Log.d(TAG, "no need to update file : " + codecWhitelistXMLFile.getName());
        }
    }

    String prepareCodecWhitelistXmlStr() {
        String[] mime_types = {"video/avc", "video/hevc"};

        ArrayList<MediaCodecInfo> decoders = new ArrayList<>();
        ArrayList<MediaCodecInfo> encoders = new ArrayList<>();

        for (String mime_type : mime_types) {
            decoders.addAll(findMediaCodecs(mime_type, false));
        }

        for (String mime_type : mime_types) {
            encoders.addAll(findMediaCodecs(mime_type, true));
        }

        return CodeLite2XmlStr(encoders, decoders);
    }

    private ArrayList<MediaCodecInfo> findMediaCodecs(String mimeType, boolean isEncoder) {
        ArrayList<MediaCodecInfo> codecList = new ArrayList<>();
        MediaCodecList list = new MediaCodecList(MediaCodecList.ALL_CODECS);
        MediaCodecInfo[] codecInfos = list.getCodecInfos();
        for (MediaCodecInfo codecInfo : codecInfos) {
            if (codecInfo.isEncoder() == isEncoder) {
                String[] types = codecInfo.getSupportedTypes();
                for (String type : types) {
                    if (type.equalsIgnoreCase(mimeType)) {
                        codecList.add(codecInfo);
                        break;
                    }
                }
            }
        }
        if (codecList.isEmpty()) {
            String message = isEncoder ? "Encoder " : "Decoder ";
            message = message + mimeType + " is not supported!";
            Toast.makeText(this, message, Toast.LENGTH_LONG).show();
        }
        return codecList;
    }

    String CodeLite2XmlStr(ArrayList<MediaCodecInfo> encoderList, ArrayList<MediaCodecInfo> decoderList) {
        /* Creates mediacodec.xml content for list of codecs supplied. The XML format is based on format specified at
         * https://github.com/open-webrtc-toolkit/owt-client-android/blob/f2294c55f9d3bdc1de78ab84c9b7d018c3e3a04b/docs/index.md?plain=1#L71.
         */

        StringBuilder xmlStr = new StringBuilder();

        xmlStr.append("<pre>\n")
                .append("  <MediaCodecs>\n")
                .append("    <Encoders>\n");

        // add encoders
        for (int i = 0; i < encoderList.size(); i++) {
            Log.d(TAG, "Encoders = " + encoderList.get(i));
            xmlStr.append("      <MediaCodec name=\"")
                    .append(encoderList.get(i).getName())
                    .append("\" type=\"")
                    .append(encoderList.get(i).getSupportedTypes()[0])
                    .append("\"/>\n");
        }

        xmlStr.append("    </Encoders>\n")
                .append("    <Decoders>\n");

        // add decoders
        for (int i = 0; i < decoderList.size(); i++) {
            Log.d(TAG, "decoders = " + decoderList.get(i));
            xmlStr.append("      <MediaCodec name=\"")
                    .append(decoderList.get(i).getName())
                    .append("\" type=\"")
                    .append(decoderList.get(i).getSupportedTypes()[0])
                    .append("\"/>\n");
        }

        xmlStr.append("    </Decoders>\n")
                .append("  </MediaCodecs>\n")
                .append("</pre>\n");

        return xmlStr.toString();
    }

}

File: aic-application/app/src/main/java/com/intel/gamepad/activity/GameDetailActivity.kt
package com.intel.gamepad.activity

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.android.simplehttpurlconnectiondemo.SimpleHttpUtils
import com.intel.gamepad.R
import com.intel.gamepad.app.*
import com.intel.gamepad.app.MyApp.context
import com.intel.gamepad.bean.GameListBean
import com.intel.gamepad.bean.GameStatusBean
import com.intel.gamepad.controller.webrtc.*
import com.intel.gamepad.fragment.RatingScoreDialogFragment
import com.intel.gamepad.fragment.RoomDialogFragment
import com.intel.gamepad.owt.p2p.P2PHelper
import com.intel.gamepad.utils.IPUtils
import com.jeremyliao.liveeventbus.LiveEventBus
import com.lzy.okgo.OkGo
import com.lzy.okgo.callback.StringCallback
import com.lzy.okgo.model.Response
import com.mcxtzhang.commonadapter.rv.CommonAdapter
import com.mcxtzhang.commonadapter.rv.ViewHolder
import com.mycommonlibrary.utils.DensityUtils
import com.mycommonlibrary.utils.LogEx
import com.mycommonlibrary.utils.StatusBarUtil
import com.mycommonlibrary.view.loadingDialog.LoadingDialog
import kotlinx.android.synthetic.main.activity_game_detail.*
import kotlinx.android.synthetic.main.activity_setting.*
import kotlinx.android.synthetic.main.include_game_cover_header.*
import kotlinx.android.synthetic.main.include_title_with_statusbar.*
import kotlinx.android.synthetic.main.item_game_image.view.*
import kotlinx.coroutines.*
import org.gaminganywhere.gaclient.PlayGameRtspActivity
import org.jetbrains.anko.longToast
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.support.v4.startActivity
import org.jetbrains.anko.toast

class GameDetailActivity : BaseActvitiy(), CoroutineScope by MainScope() {
    companion object {
        const val PARAM_BEAN = "param_bean"
        fun actionFragment(frg: Fragment, bean: GameListBean) {
            frg.startActivity<GameDetailActivity>(PARAM_BEAN to bean)
        }

        fun actionStart(act: Activity) {
            act.startActivity<GameDetailActivity>(PARAM_BEAN to GameListBean().apply {
                iid = 1
                conf = "rts"
                this.imageUrl = ""
            })
        }
    }

    private var useWebRTC = true
    private var bean: GameListBean? = null
    private var statusBean: GameStatusBean? = null
    private var repeatTimeOut = 5
    private var beginMillis = 0L
    private var nodeIP = "" // 记录需要关闭的节点IP
    private var roomCount = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        StatusBarUtil.setTranslucentStatus(this)
        setContentView(R.layout.activity_game_detail)
        initView()
        initRoomNoEvent()
        loadData()


        //Set the Android controller as the default one.
        chkAndroid.setChecked(true)
        etServerIP.setText(IPUtils.loadIP())
        P2PHelper.serverIP = IPUtils.loadIP()
        etServerIP.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                s?.let {
                    IPUtils.saveip(it.toString())
                    P2PHelper.serverIP = IPUtils.loadIP()
                }
            }
        })


        etPeerID.setText(IPUtils.loadPeerID())
        P2PHelper.peerId = IPUtils.loadPeerID()
        etPeerID.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                s?.let {
                    IPUtils.savepeerid(it.toString())
                    P2PHelper.peerId = IPUtils.loadPeerID()
                }
            }
        })

        etClientID.setText(IPUtils.loadTokenID())
        P2PHelper.clientId= IPUtils.loadTokenID()
        etClientID.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                s?.let {
                    IPUtils.savetoken(it.toString())
                    P2PHelper.clientId = IPUtils.loadTokenID()
                }
            }
        })
    }

    override fun onDestroy() {
        super.onDestroy()
        cancel()
    }

    /**
     * 在多人游戏时接收由房间列表对话框发来的房间现有人数
     * 根据这个人数选择ga?和client?
     */
    private fun initRoomNoEvent() {
        LiveEventBus.get("roomCount", Int::class.java)
            .observe(this, Observer {
                roomCount = it
                LogEx.i(">>>>>$roomCount")
            })
    }

    override fun onStart() {
        super.onStart()
        bean?.let { updateGameLog(it.iid) }
    }

    private fun loadData() {
        bean = intent.getParcelableExtra(PARAM_BEAN)

        bean?.let {
           // updateHeaderDetail(it.imageUrl)
            tvTitle.text = it.title
            //tvIntro.text = it.intro
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        LogEx.i("$requestCode $resultCode $data")
        if (requestCode == PlayGameRtspActivity.REQUEST_GAME && resultCode == Activity.RESULT_OK && data != null) {
            LogEx.i("result:" + data.getIntExtra(PlayGameRtspActivity.RESULT_MSG, -1))

            val runTime = System.currentTimeMillis() - beginMillis
            GameLogHelper.saveSingleTime(bean?.iid ?: -1, runTime)
            GameLogHelper.saveTotalTime(bean?.iid ?: -1, runTime)
            LogEx.i("run time long:$runTime")

            when (data.getIntExtra(PlayGameRtspActivity.RESULT_MSG, -1)) {
                AppConst.EXIT_NORMAL -> {
                    LogEx.i("normal exit")
                    // 正常退出时关闭游戏，并显示打分界面
                    requestCloseGame()
                    //showRatingScoreDialog()
                }
                AppConst.EXIT_TIMEOUT -> {
                    LogEx.e("time out")
                    // 网络超时或断线时重复进入游戏界面
                    if (repeatTimeOut >= 0) {
                        repeatTimeOut--
                        requestGetGameStatus()
                    } else {
                        toast("该游戏暂时无法运行，建议尝试其它游戏")
                    }
                }
                AppConst.EXIT_NOHOST -> {
                    LogEx.e("no host")
                    toast("RTSP连接HOST时出错")
                }
                AppConst.EXIT_DISCONNECT -> {
                    toast("服务器连接已断开")
                }
            }
        }
    }

    private fun initView() {
        initBackButton(R.id.ibtnBack)
        //initRvGameImage()
        // 显示创建游戏房间对话框
        /* btnJoin.setOnClickListener {
            showGameRoom()
        } */
        // 跳游戏播放界面
        btnPlay.setOnClickListener {
            requestStartGame()
        }
        btnPlay.requestFocus()
    }
/*
    private fun initRvGameImage() {
        val listGameImage = assets.list("img")?.toList()?.map { "file:///android_asset/img/$it" }

        rvGameImage.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        rvGameImage.horItemDecoration(Color.TRANSPARENT, 6f)
        rvGameImage.adapter =
            object : CommonAdapter<String>(
                context,
                listGameImage,
                R.layout.item_game_image
            ) {
                override fun convert(vh: ViewHolder, imgUrl: String) {
                    loadImage(vh.itemView.ivGame, imgUrl, R.drawable.icon_place_hold2)
                    vh.setOnClickListener(R.id.ivGame) {
                        ImageViewActivity.actionStart(this@GameDetailActivity, imgUrl)
                    }
                }
            }
    }*/

    private fun updateGameLog(gameId: Int) {
       // tvLastTime.text = GameLogHelper.loadLastTime(gameId)
      //  tvRunCount.text = "${GameLogHelper.loadRunCount(gameId)} 次"
      //  tvSingleTime.text = GameLogHelper.loadSingleTime(gameId)
       // tvTotalTime.text = GameLogHelper.getTotalTimeString(gameId)
    }

    private fun updateHeaderDetail(url: String) {
        // 作为背景模糊处理(参数25已是模糊的最大值)
       // this.loadImageBlur(ivHeaderBg, url)
       // this.loadImage(ivGameCover, url)
    }

    /**
     * 跳转到的游戏播放界面
     */
    private fun gotoGamePlay() {
        bean?.let {
            if (useWebRTC) {
                // 根据游戏类型选择游戏手柄的类型
                it.addurl = "fps"

                if (chkAndroid.isChecked) it.addurl = "android"

                val ctrlName = when (it.addurl) {
                    "fps" -> RTCControllerFPS.NAME // 打枪
                    "rts" -> RTCControllerMouse.NAME // mouse+key
                    "rac" -> RTCControllerRAC.NAME
                    "act" -> RTCControllerACT.NAME
                    "android" -> RTCControllerAndroid.NAME
                    else -> RTCControllerXBox.NAME
                }
                // TODO:以后要改信令服的地址
                //P2PHelper.serverIP = "http://192.168.1.244:8095"
                /*when (roomCount) {
                    0 -> {
                        P2PHelper.peerId = "ga2";P2PHelper.clientId = "client2"
                    }
                    in 1..3 -> {
                        P2PHelper.peerId = "ga${roomCount}"
                        P2PHelper.clientId = "client${roomCount}"
                    }
                }
                */
                LogEx.i("ga = ${P2PHelper.peerId} client = ${P2PHelper.clientId}")


                PlayGameRtcActivity.actionStart(this, ctrlName, it.iid, it.conf)
            } else {
                PlayGameRtspActivity.actionStart(this, it.iid, it.ip, it.port)
            }

            GameLogHelper.saveLastTime(it.iid)
            GameLogHelper.saveRunCount(it.iid)
            beginMillis = System.currentTimeMillis()
        }
    }

    private fun showRatingScoreDialog() {
        RatingScoreDialogFragment.newInstance().show(supportFragmentManager, "dialog")
    }

    /**
     * 通知服务器关游戏
     */
    private fun requestCloseGame() {

        P2PHelper.closeP2PClient();
       // longToast("turn off the game")
        /*if (nodeIP.isEmpty()) return
        OkGo.get<String>(IPUtils.loadIP() + AppConst.CLOSE_GAME)
            .params("iid", bean?.iid.toString())
            .params("gamedir", MyApp.pId)
            .params("nodeIP", nodeIP)
            .execute(object : StringCallback() {
                override fun onSuccess(response: Response<String>) {
                    LogEx.i(response.body())
                    nodeIP = ""
                }
            })
        */
    }

    /**
     * 获取游戏的状态，用于判断是否可以玩游戏
     */
    private fun requestGetGameStatus() {
        val dlg = LoadingDialog.show(this)
        OkGo.get<String>(IPUtils.loadIP() + AppConst.GAME_STATUS)
            .params("iid", bean?.iid.toString())
            .params("gamedir", MyApp.pId)
            .execute(object : StringCallback() {
                override fun onSuccess(response: Response<String>) {
                    LogEx.i(response.body())
                    val result = GameStatusBean.objectFromData(response.body())
                    if (result.status == 0) {
                        statusBean = result
                    } else {
                        longToast("该游戏已被占用，请选择别的游戏！")
                    }
                }

                override fun onError(response: Response<String>) {
                    super.onError(response)
                    LogEx.e(response.exception.message)
                    longToast("连接服务器失败！请检查网络是否正常！")
                }

                override fun onFinish() {
                    super.onFinish()
                    dlg?.dismiss()
                }
            })
    }

    /**
     * 加入游戏（多人玩同一个节点的游戏）
     */
    fun requestJoinGame() = launch {
        val dlg = LoadingDialog.show(this@GameDetailActivity)
        withContext(Dispatchers.IO) {
            // 获取游戏节点IP
            nodeIP = SimpleHttpUtils.get(IPUtils.loadIP() + AppConst.JOIN_GAME, emptyMap())
        }
        dlg.dismiss()
        P2PHelper.serverIP = "http://${nodeIP}:8095/"
        gotoGamePlay() // 跳游戏播放界面
    }

    /**
     * 删除游戏房间
     * 只能删除自己创建的房间，而且这个操作最好交由服务端完成，以防断网时无法删除
     */
    fun requestDeleteRoom() = launch(Dispatchers.IO) {
        OkGo.get<String>(IPUtils.loadIP() + AppConst.DEL_ROOM)
            .params("roomname", UserHelper.getUserName())
            .execute()
    }

    /**
     * 创建游戏房间
     * 为防止房间刚创建完成游戏就被别人占了的情况，应该先启动游戏再创建房间
     */
    private fun requestCreateRoom() = launch(Dispatchers.IO) {
        val response = OkGo.post<String>(IPUtils.loadIP() + AppConst.CREATE_ROOM)
            .tag(this)
            .params("roomname", UserHelper.getUserName())
            .params("gameid", bean?.iid ?: 0)
            .params("roomtotalcount", "1234")
            .params("creatroomuserid", UserHelper.getId())
            .params("gameip", nodeIP)  // 游戏的节点IP
            .execute()?.body?.string()
        LogEx.i(response)
    }

    /**
     * 启动游戏
     * @param isCreateRoom 是否在启动成功后创建房间
     */
    fun requestStartGame(isCreateRoom: Boolean = false) {
        val dlg = LoadingDialog.show(this)
        if(isCreateRoom){
            longToast("room is not supported");
        }
        else {
            /* OkGo.get<String>(IPUtils.load() + AppConst.START_GAME)
            .params("iid", bean?.iid ?: 0)
            .params("gamedir", MyApp.pId) // 唯一码
            .params("width", DensityUtils.getmScreenHeight())
            .params("height", DensityUtils.getmScreenWidth())
            .params("fps", 60)
            .params("client", 1)
            .execute(object : StringCallback() {
                override fun onSuccess(response: Response<String>) {
                    LogEx.i(response.body())
                    //val result = GameStatusBean.objectFromData(response.body())
                    //if (result.isSuccess) {
                       // nodeIP = result.ip
                        P2PHelper.serverIP = "http://10.188.197.40:8095/"
                        LogEx.i(">>>> ${P2PHelper.serverIP}")
                        gotoGamePlay()

                       // if (isCreateRoom) requestCreateRoom()

                    //} else {
                        //longToast("" + result.message)
                    //}
                }

                override fun onError(response: Response<String>?) {
                    super.onError(response)
                    dlg.dismiss()
                    response?.exception?.printStackTrace()
                    LogEx.e("Error: " + response?.exception?.message)
                    toast(response?.exception?.message.toString())
                }

                override fun onFinish() {
                    super.onFinish()
                    dlg.dismiss()
                }
            })*/
            gotoGamePlay()
            dlg.dismiss()
        }
    }

    /**
     * 显示游戏房间列表对话框
     * 多人模式，可以创建房间，可以加入房间
     */
    private fun showGameRoom() {
        val iid = bean?.iid ?: 0
        RoomDialogFragment.newInstance(iid).show(supportFragmentManager, "dialog")
    }
}
